### PR TITLE
Fix: Adding global middlewares, services and utils

### PR DESCRIPTION
**Pull request by AngeloChristian1**
#### What does this PR do?

- Added middlewares for Authentication, isOwner and checkPermision
- Added Utils for password hashing, token encoding and decoding
- Added user services

### CODE REVIEW

File: .vscode/settings.json
Comment (munezerobagira) [inline .vscode/settings.json:1]: What is the purpose of this configurations?

Comment (munezerobagira) [reply .vscode/settings.json:1]: It would be better if you add .vscode/settings.json in .gitignore

File: src/controllers/LoginController.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+   * @param {Request} req - Express request object
+   * @param {Response} res - Express response object
+   * @returns {Promise<Response>} Promise that resolves to an Express response
+   */
+  static async login(req: Request, res: Response): Promise<Response> {
+    // Validate request body against login schema
+    const { error } = loginSchema.validate(req.body)
+    if (error) {
+      return res.status(400).json({ message: error.details[0].message })
+    }
```
Comment (munezerobagira) [inline src/controllers/LoginController.ts]: why this seem like new file for login controller, is this new feature?

File: src/database/migrations/20240427140502-create-user.js
Comment (munezerobagira) [inline src/database/migrations/20240427140502-create-user.js:1]: don't delete migration

File: src/docs/loginDocs.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ *               password:
+ *                 type: string
+ *     responses:
+ *       '200':
+ *         description: Successful login, returns a JWT token
+ *         content:
+ *           application/json:
+ *             schema:
+ *               type: object
+ *               properties:
```
Comment (munezerobagira) [inline src/docs/loginDocs.ts]: why do we have new swagger documentation, is this new feature too?

File: src/middlewares/index.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        const existingUser = await getUserById(userInfo?.id)
+
+        if(!existingUser){
+            return res.status(401).json({message:"Unauthorized"});
+        }
+        if(existingUser.id as number !== Number(id)){
+            return res.status(401).json({message:"Unauthorized"})
+        }
+            next()
+
```
Comment (munezerobagira) [inline src/middlewares/index.ts]:
```suggestion

        
```
this would be implemented #49 , this middleware doesn't have be included in this fix

**Review by munezerobagira — CHANGES_REQUESTED** [review]

File: src/config/index.ts [hunk 4]
```diff
@@ -7,24 +7,24 @@ export const DOCKER_LOCAL_PORT = parseInt(process.env.DOCKER_LOCAL_PORT, 10)
...
-export const TEST_DATABASE_URL = process.env.TEST_DATABASE_URL
-export const PROD_DATABASE_URL = process.env.PROD_DATABASE_URL
-export const GOOGLE_CLIENT_ID = process.env.GOOGLE_CLIENT_ID
-export const GOOGLE_CLIENT_SECRET = process.env.GOOGLE_CLIENT_SECRET
+export const { DB_HOST_TYPE } = process.env
+export const { DEV_DATABASE_URL } = process.env
+export const { TEST_DATABASE_URL } = process.env
+export const { PROD_DATABASE_URL } = process.env
+export const { GOOGLE_CLIENT_ID } = process.env
+export const { GOOGLE_CLIENT_SECRET } = process.env
```
Comment (munezerobagira) [inline src/config/index.ts]:
```suggestion
export const { DB_HOST_TYPE, DEV_DATABASE_URL, TEST_DATABASE_URL, PROD_DATABASE_URL, GOOGLE_CLIENT_ID, GOOGLE_CLIENT_SECRET} = process.env
```
this can be combined into single export

File: src/utils/sendEmail.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  try {
+    const transporter = nodemailer.createTransport({
+      service: process.env.NODEMAILER_SERVICE,
+      host: process.env.NODEMAILER_HOST,
+      port: parseInt(process.env.NODEMAILER_EMAIL_PORT || '', 10),
+      secure: process.env.SECURE === 'true',
+      auth: {
+        user: process.env.NODEMAILER_USER,
+        pass: process.env.NODEMAILER_PASS,
+      },
```
Comment (munezerobagira) [inline src/utils/sendEmail.ts]:
```suggestion
import {NODEMAILER_SERVICE,NODEMAILER_HOST, NODEMAILER_SECURE, NODEMAILER_PASS} from "../config"
      service: NODEMAILER_SERVICE,
      host: NODEMAILER_HOST,
      port: NODEMAILER_EMAIL_PORT,
      secure:NODEMAILER_SECURE === 'true',
      auth: {
        user: NODEMAILER_USER,
        pass: NODEMAILER_PASS,
      },
```

File: src/middlewares/authenticationMiddleware.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      process.env.JWT_SECRET || 'USER-AUTH',
+      (err: unknown, decoded: unknown) => {
+        if (err) {
+          return res.status(401).json({ message: 'Invalid token' })
+        }
+        res.locals.decoded = decoded
+        req.userDetails = decoded
+        next()
+      },
+    )
```
Comment (munezerobagira) [inline src/middlewares/authenticationMiddleware.ts]:
```suggestion
   const payload=await decodeToken(token)
   const user=await getUserById(payload.id)
   req.user=user;
   res.locals.decoded
   next()

```
We can use JWT utils instead of hardcording jwt verfication and also  convert this async await for blocking, to make sure that each handler have req.user

File: src/utils/jwtConfiguration.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import jwt from 'jsonwebtoken'
+
+export const generateToken = (data: object) => {
+  const token = jwt.sign(data, process.env.JWT_SECRET || 'USER-AUTH', {
```
Comment (munezerobagira) [inline src/utils/jwtConfiguration.ts]:
```suggestion
  const token = jwt.sign(data, JWT_SECRET, {
```
its better to import enviroment variables from `src/config`

File: src/utils/jwtConfiguration.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    expiresIn: process.env.JWT_EXPIRE_TIME || 300000,
```
Comment (munezerobagira) [inline src/utils/jwtConfiguration.ts]:
```suggestion
    expiresIn: JWT_EXPIRE_TIME,
```
its better to import environment  from `src/config`

File: src/utils/jwtConfiguration.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  })
+  return token
+}
+
+export const decodeToken = (token: string) => {
+  const verify = jwt.verify(token, process.env.JWT_SECRET || 'USER-AUTH')
```
Comment (munezerobagira) [inline src/utils/jwtConfiguration.ts]:
```suggestion
  const verify = jwt.verify(token, JWT_SECRET)
```
we can import this environment from config too

File: src/utils/passwords.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import bcrypt from 'bcrypt'
+import dotenv from 'dotenv'
+
+dotenv.config()
+
+export const hashPassword = async (password: string) => {
+  const slatRounds = await bcrypt.genSalt(
+    parseInt(process.env.SLAT_ROUND, 10) || 10,
```
Comment (munezerobagira) [inline src/utils/passwords.ts]:
```suggestion
    SLAT_ROUND,
```
you can export from `src/config` and use it and import it here

File: src/utils/sendEmail.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    })
+
+    transporter.sendMail({
+      from: process.env.NODEMAILER_USER,
```
Comment (munezerobagira) [inline src/utils/sendEmail.ts:25]:
```suggestion
      from: NODEMAILER_USER,
```

File: src/config/index.ts [hunk 12]
```diff
@@ -7,24 +7,24 @@ export const DOCKER_LOCAL_PORT = parseInt(process.env.DOCKER_LOCAL_PORT, 10)
...
-export const NODEMAILER_USER = process.env.NODEMAILER_USER
-export const NODEMAILER_PASS = process.env.NODEMAILER_PASS
-export const NODEMAILER_BASE_URL = process.env.NODEMAILER_BASE_URL
-export const REDIS_PASSWORD = process.env.REDIS_PASSWORD
-export const REDIS_HOST = process.env.REDIS_HOST
+export const { NODEMAILER_USER } = process.env
+export const { NODEMAILER_PASS } = process.env
+export const { NODEMAILER_BASE_URL } = process.env
+export const { REDIS_PASSWORD } = process.env
+export const { REDIS_HOST } = process.env
```
Comment (munezerobagira) [inline src/config/index.ts]:
```suggestion
``` 
these also can be combined into single export

File: src/services/userServices.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+export const createUser = async (details: object) => {
+  const user = await User.create(details)
+  return user
+}
+
+export const getUserByRole = (userRole: string) =>
+  User.findOne({ where: { userRole } })
+
+export const deleteUserById = (id: string) => User.destroy({ where: { id } })
```
Comment (munezerobagira) [inline src/services/userServices.ts]: good work on adding service

File: src/middlewares/authenticationMiddleware.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    const token = req.headers.authorization?.split(' ')[1] || ''
+
+    if (!token) {
+      return res.status(401).json({ message: 'no access token found' })
+    }
+
+    jwt.verify(
+      token,
+      process.env.JWT_SECRET || 'USER-AUTH',
```
Comment (munezerobagira) [inline src/middlewares/authenticationMiddleware.ts]:
```suggestion
      JWT_SECRET
```
JWT_SECRET can be imported from the `src/config`, so th dat whenever name change  for env you don't have change it multiple place

File: src/database/migrations/20240427140502-create-user.js
Comment (munezerobagira) [inline src/database/migrations/20240427140502-create-user.js:1]: why did you delete this migration

File: src/utils/jwtConfiguration.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+  return verify
```
Comment (munezerobagira) [inline src/utils/jwtConfiguration.ts]:
```suggestion
  return  jwt.verify(token, JWT_SECRET )
```
can return jwt.verify directly without needing verify as it not being used

File: src/middlewares/authenticationMiddleware.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+  namespace Express {
+    // eslint-disable-next-line no-shadow
+    interface Request {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      userDetails: any
+    }
+  }
+}
+
+// eslint-disable-next-line import/prefer-default-export
```
Comment (munezerobagira) [inline src/middlewares/authenticationMiddleware.ts]:
```suggestion
declare global {
  namespace Express {
    interface Request {
      userDetails: any
    }
  }
}

```
Lets not disable eslint

**Review by munezerobagira — CHANGES_REQUESTED** [review]
Good work but there are few things which we can improve

**Comment by munezerobagira** [comment]
It seem that fix PR includes alot of fix, if possible we can break it down into small fixes

File: src/middlewares/authenticationMiddleware.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable @typescript-eslint/no-namespace */
+/* eslint-disable no-shadow */
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (munezerobagira) [inline src/middlewares/authenticationMiddleware.ts]:
```suggestion

```

File: src/services/userServices.ts [hunk 18]
```diff
@@ -0,0 +1,27 @@
+import User from '../database/models/userModel'
+
+export const getUserByEmail = async (email: string) => {
+  const user = await User.findOne({ where: { email } })
+  return user
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (munezerobagira) [inline src/services/userServices.ts]:
```suggestion

```

File: src/services/userServices.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+export const getUserByUsername = async (username: string) => {
+  const user = await User.findOne({ where: { username } })
+  return user
+}
+
+export const createUser = async (details: object) => {
+  const user = await User.create(details)
+  return user
```
Comment (munezerobagira) [inline src/services/userServices.ts]:
```suggestion
  return  User.create(details)
 
```

File: src/utils/jwtConfiguration.ts [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+import { JWT_SECRET, JWT_EXPIRE_TIME } from '../config/index'
+
+export const generateToken = (data: object) => {
+  const token = jwt.sign(data, JWT_SECRET, {
+    expiresIn: JWT_EXPIRE_TIME,
+  })
+  return token
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (munezerobagira) [inline src/utils/jwtConfiguration.ts]:
```suggestion
```

File: src/utils/jwt.ts
Comment (munezerobagira) [inline src/utils/jwt.ts:1]: why do we have two jwt utils?

File: src/middlewares/authenticationMiddleware.ts [hunk 21]
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable @typescript-eslint/no-namespace */
+/* eslint-disable no-shadow */
```
Comment (munezerobagira) [inline src/middlewares/authenticationMiddleware.ts]:
```suggestion

```

**Review by munezerobagira — CHANGES_REQUESTED** [review]

File: src/utils/jwtUtils.ts [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+  const token = jwt.sign(data, JWT_SECRET, {
+    expiresIn: JWT_EXPIRE_TIME,
+  })
+  return token
+}
+
+interface JwtPayload {
+  userId: string
+  id: string
+}
```
Comment (mpanoathos) [inline src/utils/jwtUtils.ts]: I think this file was created before by Eric, Why did you add this part?

**Review by mpanoathos — CHANGES_REQUESTED** [review]

File: src/services/userServices.ts [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+
+export const getUserByUsername = async (username: string) =>
+  User.findOne({ where: { username } })
+
+export const createUser = async (details: object) => User.create(details)
+
+export const getUserByRole = (userRole: string) =>
+  User.findOne({ where: { userRole } })
+
+export const deleteUserById = (id: string) => User.destroy({ where: { id } })
```
Comment (mpanoathos) [inline src/services/userServices.ts]: Nice

File: src/services/userServices.ts [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+export const getUserByEmail = async (email: string) =>
+  User.findOne({ where: { email } })
+
+export const getUserById = async (id: number) => {
+  const user = await User.findOne({ where: { id } })
+  return user
+}
+
+export const getUserByUsername = async (username: string) =>
+  User.findOne({ where: { username } })
```
Comment (Yvan-David) [inline src/services/userServices.ts]: username is not unique please you should use findAll

File: src/services/userServices.ts [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+
+export const createUser = async (details: object) => User.create(details)
+
+export const getUserByRole = (userRole: string) =>
+  User.findOne({ where: { userRole } })
```
Comment (Yvan-David) [inline src/services/userServices.ts]: userRole is not unique please you should use findAll

**Review by Yvan-David — CHANGES_REQUESTED** [review]

File: src/utils/jwtUtils.ts (hunk 22 above)
Comment (AngeloChristian1) [reply src/utils/jwtUtils.ts]: I created a datatype for JWT token decoded content

File: src/config/index.ts [hunk 26]
```diff
@@ -1,19 +1,20 @@
+/* eslint-disable prefer-destructuring */
```
Comment (munezerobagira) [inline src/config/index.ts]:
```suggestion
```
I guess that this is no longer needed

File: src/middlewares/authenticationMiddleware.ts [hunk 27]
```diff
@@ -0,0 +1,34 @@
+import { Response, Request, NextFunction } from 'express'
+import { decodeToken } from '../utils/jwt'
+import { getUserById } from '../services/userServices'
+
+export interface MyExtendedRequest extends Request {
+  userDetails?: object
+}
```
Comment (munezerobagira) [inline src/middlewares/authenticationMiddleware.ts]:
```suggestion
export interface MyExtendedRequest extends Request {
  userDetails?: Model<UserAttributes>
}
```
This can be done through .d.ts file so that it can be globally recognized
https://stackoverflow.com/questions/37377731/extend-express-request-object-using-typescript
and instead of calling it userDetails, you can call it, user and make it simple

**Review by munezerobagira — CHANGES_REQUESTED** [review]

File: src/middlewares/authenticationMiddleware.ts (hunk 27 above)
Comment (AngeloChristian1) [reply src/middlewares/authenticationMiddleware.ts]: The file `express.d.ts` wasn't working so I did it directly in this file

declare global {
  namespace Express {
    interface Request {
      user?: object
    }
  }
}

**Review by Yvan-David — APPROVED** [review]
Great Job

**Review by mpanoathos — APPROVED** [review]

**Review by munezerobagira — APPROVED** [review]
